### DISCUSSION TITLE
I did lingui compile but in production the translations are showing only the ids of keys, it is not using defaultLocale translation.

### DISCUSSION THREAD

**Discussion by crazyyi** [lingui/js-lingui#1518] [Q&A]
It is working fine in dev. So I have this config:

In `i18n.ts` I have this:
```javascript
export function activateLocale(i18n: I18n, locale: string, messages: Messages) {
  if (locale === "pseudo") {
    locale = "en";
  }

  const cardinalRules = new Intl.PluralRules(locale, { type: "cardinal" });
  const ordinalRules = new Intl.PluralRules(locale, { type: "ordinal" });

  i18n.loadLocaleData(locale, {
    plurals: (n, ordinal) => {
      return (ordinal ? ordinalRules : cardinalRules).select(n);
    },
  });

  i18n.load(locale, messages);
  i18n.activate(locale);
}
```
and in `next.config.js` I have this:
```javascript
i18n: {
    locales: ['en', 'ko', 'zh-Hant', 'zh', 'es', 'fr', 'ms', 'vi', 'ja', 'pseudo'],
    defaultLocale: 'en'
  },
```
`lingui.config.js` is like this:
```javascript
const nextConfig = require("./next.config");
module.exports = {
  locales: nextConfig.i18n.locales,
  pseudoLocale: "pseudo",
  sourceLocale: nextConfig.i18n.defaultLocale,
  fallbackLocales: {
    default: nextConfig.i18n.defaultLocale,
  },
  extractBabelOptions: {
    presets: ["next/babel"]
  },
  catalogs: [
    {
      path: "src/locales/{locale}/message",
      include: ["src/pages", "src/components"],
    },
  ],
  format: "minimal",
};

```
Anything I am missing in the config? Do I need ot explicitly config the `fallbacklng`?

**Comment by timofei-iatsenko** [comment]
First, please specify a syntax in your code snippets, so github would apply a syntax highlighting for them https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code-and-syntax-highlighting

Second, the input provided by you is not enough to say what is going wrong. I don't see how catalogs are loaded in the app. and how do you compile them. 

Couple suggestion though:

- don't use `minimal`, or `lingui` or any other json type of catalogs. They are not created for that. Use `po` catalogs, they much better and all tooling in lingui assuming you use `po` so all integration would be easier
-  try to use v4, there are many improvements related to fallbacks

**Reply by crazyyi** [reply]
@thekip 

Thanks. I have updated the syntax highlights.
I use minimal because the translators/PMs requested "json" format. They might be confused about the "po" format.

For loading catalogs I have this: (in `i18n.ts`)
```javascript
export async function loadTranslation(locale: string, filename: string) {
  const messages = await import(`./locales/${locale}/${filename}`);
  return messages;
}
```

And in _app.tsx I put this:
```javascript
if (pageProps.translation && firstRender.current) {
    activateLocale(i18n, locale, pageProps.translation)
    firstRender.current = false
  }

  useEffect(() => {
    if (pageProps.translation) {
      activateLocale(i18n, locale, pageProps.translation)
    }
  }, [locale, pageProps.translation])
```

And here is a typical getServerSideProps for SSR:
```javascript
export const getServerSideProps = async (context: any) => {
  const translation = await loadTranslation(context.locale!, "faq");

  return {
    props: {
      isDev: process.env.NODE_ENV === "development",
      translation: { ...translation },
    },
  };
};
```

I will try v4 if all attempts don't work.

**Reply by timofei-iatsenko** [reply]
```ts
export async function loadTranslation(locale: string, filename: string) {
  const messages = await import(`./locales/${locale}/${filename}`);
  return messages;
}
```

Are you sure here is **compiled** catalogs loaded? The extension is not specified, it might load your "minimal" json instead of compiled.

**Reply by crazyyi** [reply]
> ```ts
> export async function loadTranslation(locale: string, filename: string) {
>   const messages = await import(`./locales/${locale}/${filename}`);
>   return messages;
> }
> ```
> 
> Are you sure here is **compiled** catalogs loaded? The extension is not specified, it might load your "minimal" json instead of compiled.

Sorry I was generalizing here. In the actual catalogs config I did provide all the extensions (such as "tsx") in the `include` array, and for folders I only have the "index.tsx" files inside. I am separating the translations into different files so I am not posting all of them here.
I thought the final part of the import is just a filename of the `json` file. But in `production` it seems to be the js file and the import will return the messages object.
So how to config it to load `json` file in dev and load `js` file in `production`?

**Reply by timofei-iatsenko** [reply]
Your import code is ambiguous, i assume you have compiled catalogs (.js) next to your sources (.json) and from this import it's not clear which file webpack would consume. I suspect that it's picking up a json which is not compiled. Try this 

```js
export async function loadTranslation(locale: string, filename: string) {
  const messages = await import(`./locales/${locale}/${filename}.js`);
  return messages;
}
```

Also check that this  `import` does not return a `default`  

```js
export async function loadTranslation(locale: string, filename: string) {
  const {messages: default} = await import(`./locales/${locale}/${filename}.js`);
  return messages;
}
```

And in general check / show here what is inside you compiled catalogs. It should have all your keys with a translations.

**Reply by crazyyi** [reply]
@thekip 
Yes I have both the js file and the `json` file and both of them have keys and translations.
so we can skip the "json" file in `loadTranslation` totally? In dev which file is `lingui` picking up?
I saw that in the provided example it's using `messages.po` so I am not sure how it deals with json files.

Also in `typescript` I use `default` but got this error:
`Identifier expected. 'default' is a reserved word that cannot be used here.`

**Reply by timofei-iatsenko** [reply]
If you build pipeline uses webpack (nextjs do use) you can leverage `@lingui/loader` as in the example.

The common misunderstanding of people is related to json catalogs. People thinks they can consume it directly in the app and compile step is needed only for `po` files, which is not natively supported by js ecosystem.

The truth is, you should always compile your catalogs, no matter prod or dev. And this is very annoying. That's why `@lingui/loader`  exists (and vite + snowpack plugins as well). 

So raw catalogs (json, po) are used only for transferring between you and your translator. In app you should always use compiled version. 

You can completely omit a separate step of compiling catalogs and use loader as shown in the example:

```ts
const messages = (await import(`@lingui/loader!./locales/${locale}/${filename}.json`)).default;
```

> I use minimal because the translators/PMs requested "json" format.

For sure, your translators know what to do with po files. It's industry standard for i18n since 1990 (33 years!) They just said first thing came in to mind.

**Reply by crazyyi** [reply]
@thekip 
Thank you.
After some experiment, I finally figure out how to use `@lingui/loader`.
I can only use it with the PO format, like this:
```javascript
const messages = (await import(`@lingui/loader!./locales/${locale}/${filename}.po`)).default;
```
I can never get it to work with `json` files. It would always give an error saying "Unexpected token in JSON at position 0" parsing every single one of the `json` file. 
Looks like the loader is having some problem parsing json files. Maybe I'll need to use a custom loader for that purpose which I don't know how to start.
Anyway, this suggested solution would end up with fewer files in the code base and fewer steps because the loader would do the compile work for me.

I will see if I can convince them to use the PO format. The PMs a.k.a. "translators" have little knowledge of i18n. They know spreadsheets better than anything else. 

I gained a lot of knowledge from modifying an existing example on codesandbox, which is similar to my setup. Here is the link to the modified small example.
https://codesandbox.io/s/lingui-nextjs-typescript-example-mm9l7m

**Reply by timofei-iatsenko** [reply]
> I will see if I can convince them to use the PO format. The PMs a.k.a. "translators" have little knowledge of i18n. They know spreadsheets better than anything else.

Just show them PoEditor. They don't have to change them by hands. But even by hands it's far more readable than json (which has much stricter grammar)

> I can never get it to work with json files. I

Indeed, it has never tested with the json files. But at least from the code point of view it should work. You can create an issue

**Reply by crazyyi** [reply]
@thekip 
Thanks. PoEditor looks awesome.

In the [docs](https://lingui.dev/ref/loader), the example is showing using the loader for JSON files. But I think it should change to PO if the loader hasn't been tested on JSON files.

Since we are using `@lingui/loader`, which is called in `loadTranslation` function which in Nextjs is called at `getServerSideProps` during SSR, are the languages loaded at runtime or at compile time? If it's at runtime would it slow the performance?

There are still some issues displaying the translations in production mode. `yarn dev` and `yarn prod` yield different results. 
In dev mode, I can see all the default translations which is in English. But in production, all I can see are the keys of the texts. I am already using `@lingui/loader` to load `po` format locale files.
I put a log after `loadTranslation` in `getStsaticProps`.  `loadTranslation` basically just return 
```javascript
(await import(`@lingui/loader!./locales/${locale}/${filename}.po`)).default; 
```
In `yarn build` I can see all the translations are ready. And the `i18n` has the correct **messages** object attached to it which contains all the default translations. But I just can't see the translations in the production page.
I have tried using `i18n._({id: "", message: ""})` directly without macros and it works in production. Both `t` and `Trans` macros are not working in production. Not sure what is going wrong.

I found another package in the project `@lingui/remote-loader`, I tried that to load my JSON files but seems it's not compatible with SSR. There is a prerender-error saying 
```
TypeError: Cannot read properties of undefined (reading 'translation')
```

**Reply by timofei-iatsenko** [reply]
Show your code.

**Reply by timofei-iatsenko** [reply]
https://github.com/thekip/nextjs-translation-demo/tree/update-for-latest-lingui compare with working example

**Reply by crazyyi** [reply]
@thekip 
Thanks again. I look at the example you wrote very carefully, and find that I am missing the bracket around `messages` variable in 
```
const messages = await import(`@lingui/loader!./locales/${locale}/${filename}.po`);
```
This mistake wouldn't throw an error during development or production and I can attach all the translations to i18n variable. It just wouldn't put the translations on the production site.

I thought dynamic import would return the translation object but it's returning an object containing the `messages` object.

FYI, this:
```
const messages = (await import(`@lingui/loader!./locales/${locale}/${filename}.po`)).default;
```
doesn't work as well. That's why my build kept failing.

Only this one works in my case:
```
const { messages } = await import(`@lingui/loader!./locales/${locale}/${filename}.po`);
```

**Reply by timofei-iatsenko** [reply]
Glad you find the issue. It all depends on your bundler / transpiler configuration. For some it would return with `default` for other return `messages` directly (depends on esModuleInterop flag).  

By the way. The reason why it works in development and doesn't work in production is that in development, Macro left in the code source messages to make DX better. But in production source messages dropped from the bundle to save bytes.

That's why in development you had fallback to English string (source) and in production you had fallback to keys (source messages was dropped and catalogs was not loaded)

**Reply by crazyyi** [reply]
@thekip 
Thank you for the explanation. It would be of great help for the users if the loader throws a warning/error when the translations are not correctly loaded.
